### PR TITLE
Update network error handling

### PR DESCRIPTION
**Pull request by khawkins**
Instituting a new workflow to preprocess errors in network responses, before further processing of their content.  Allows functionality like `SFRestAPISalesforceAction.parseResponse` to interoperate with auth token refresh functionality upstream.

### CODE REVIEW

**Comment by khawkins** [comment]
I changed up the error handling pattern a bit, to handle errors in more of a pipeline.  That way, each `CSFAction` class can add their own variation to error conditions that are important to them.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFAction.m [hunk 1]
```diff
@@ -756,24 +756,42 @@ - (id)contentFromData:(NSData*)data fromResponse:(NSHTTPURLResponse*)response er
...
-    if (jsonParseError && error && requestSucceeded) {
+    if (jsonParseError && requestSucceeded) {
         NetworkWarn(@"Error while parsing response; it doesn't appear to be JSON");
         
-        *error = [NSError errorWithDomain:CSFNetworkErrorDomain
-                                     code:CSFNetworkJSONInvalidError
-                                 userInfo:@{ NSLocalizedDescriptionKey: @"Processing response content failed",
-                                             NSUnderlyingErrorKey: jsonParseError,
-                                             CSFNetworkErrorActionKey: self }];
+        if (error) {
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFAction.m]: Logic's the same, but was hard to read in its previous context.  Also missed the warning if `error` was `nil`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Queue/CSFNetwork+Internal.h [hunk 2]
```diff
@@ -72,12 +72,6 @@
 
 - (CSFAction*)duplicateActionInFlight:(CSFAction*)action;
 
-/*!
- Provides a means for low-level code to tell us that this device has been de-authorized.
- This can happen if the device is unchecked from Chatter settings in the org setup.
- */
-- (void)receivedDevicedUnauthorizedError:(CSFAction*)action;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Queue/CSFNetwork+Internal.h]: This empty method is not used anywhere.  Removing.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m [hunk 3]
```diff
@@ -118,111 +118,127 @@ - (NSDictionary *)headersForAction {
...
+            // unavailable - too many requests in an hour
+            // max concurrent requests hit - max concurrent requests is a setting, when last confirmed it was 50 per JVM)
+            break;
+            
+        case 500:
+            // all other errors: "An error has occurred within Force.com"
+            // fall-through
+            
+        default:
+            // When the user is revoked, the response object is nil (so error code is 0)
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m]: I don't understand this case.
What's the status code for a revoke call?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m [hunk 4]
```diff
@@ -118,111 +118,127 @@ - (NSDictionary *)headersForAction {
...
+    return content;
+}
+
+- (NSError *)errorFromData:(NSData *)data response:(NSHTTPURLResponse *)response {
+    // Check for upstream errors.
+    NSError *superError = [super errorFromData:data response:response];
+    
+    if (response.statusCode < 400) {
+        // Nothing that we care about, from an error standpoint.
+        return superError;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m]: Probably...but the inherited class intentionally doesn't care—it's an implementation detail in `CSFAction`.  If that implementation changes to have an error in that case, this code stays the same.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m (hunk 3 above)
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m]: Not sure, honestly.  This code is just moved from its original location in `contentFromData:fromResponse:error:`.  I figured we could further refine this error handling, once we have a better idea of the server-side contract.  But in the meantime, it should cover whatever cases it covered before.

**Pull request merged**